File: FSUIKit/Classes/FSTabsView.m
//
//  FSTabsView.m
//  FBRetainCycleDetector
//
//  Created by FudonFuchina on 2020/1/12.
//

#import "FSTabsView.h"

@interface FSTabsView ()

@end

@implementation FSTabsView

- (void)setList:(NSArray<NSString *> *)list {
    _list = nil;
    if ([list isKindOfClass:NSArray.class] && list.count) {
        _list = list;
    }
    
    for (FSTabView *tab in self.subviews) {
        if ([tab isKindOfClass:FSTabView.class]) {
            tab.hidden = YES;
        }
    }
    
    if (!_list.count) {
        return;
    }
    
    CGFloat w = self.frame.size.width / list.count;
    for (int x = 0; x < list.count; x ++) {
        CGRect fr = CGRectMake(w * x, 0, w, self.frame.size.height);
        FSTabView *tab = [FSView viewWithTheTag:x inView:self];
        if (tab) {
            tab.hidden = NO;
            tab.frame = fr;
        } else {
            tab = [[FSTabView alloc] initWithFrame:fr];
            tab.theTag = x;
            [self addSubview:tab];
            __weak typeof(self)this = self;
            tab.click = ^(FSView * _Nonnull view) {
                if (this.clickIndex) {
                    this.clickIndex(this, view.theTag);
                }
            };
            tab.selectedState = ^(FSTabView * _Nonnull tabV, BOOL selected) {
                if (this.selectedState) {
                    this.selectedState(tabV, selected);
                }
            };
        }
        tab.label.text = list[x];
    }
}

- (void)selectedIndex:(NSInteger)index {
    for (FSTabView *tab in self.subviews) {
        if ([tab isKindOfClass:FSTabView.class]) {
            if (tab.theTag == index) {
                tab.selected = YES;
            } else {
                tab.selected = NO;
            }
        }
    }
}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

@end

File: FSUIKit/Classes/Swift/FSTabsView.swift
// FSTabsViewS.swift
// Translated from FSTabsView.h/m

import UIKit

@objcMembers
public class FSTabsView: FSView {
    
    public var list: [String]? {
        didSet {
            setupTabs()
        }
    }
    
    public var clickIndex: ((FSTabsView, Int) -> Void)?
    public var selectedState: ((FSTabView, Bool) -> Void)?
    
    private func setupTabs() {
        // Hide existing tabs
        for sub in subviews {
            if let tab = sub as? FSTabView {
                tab.isHidden = true
            }
        }
        
        guard let list = list, !list.isEmpty else { return }
        
        let w = frame.width / CGFloat(list.count)
        for (index, text) in list.enumerated() {
            let fr = CGRect(x: w * CGFloat(index), y: 0, width: w, height: frame.height)
            
            var tab: FSTabView
            if let existingTab = FSView.viewWithTheTag(tag: index, view: self) as? FSTabView {
                existingTab.isHidden = false
                existingTab.frame = fr
                tab = existingTab
            } else {
                tab = FSTabView(frame: fr)
                tab.theTag = index
                addSubview(tab)
                
                tab.click = { [weak self] view, p in
                    guard let self = self else { return }
                    if let tabView = view as? FSTabView {
                        self.clickIndex?(self, tabView.theTag)
                    }
                }
                tab.selectedState = { [weak self] tabV, selected in
                    self?.selectedState?(tabV, selected)
                }
            }
            tab.label.text = text
        }
    }
    
    public func selectedIndex(_ index: Int) {
        for sub in subviews {
            if let tab = sub as? FSTabView {
                tab.selected = (tab.theTag == index)
            }
        }
    }
}
